include/core/shape.h

```diff
@@ -133,6 +134,952 @@ class Shape : public AutoDerivativeFunction<dim>
     const typename DoFHandler<dim>::active_cell_iterator cell,
     const unsigned int                                   component = 0);
 
+  /**
+   * @brief Return the smoothed maximum of two variables used for shape contact.
+   * @param a first variable
+   * @param b second variable
+   */
+  double
+  smooth_max(double a, double b)
+  {
+    double smooth_factor = 10;
+    return (a * std::exp(a * smooth_factor) + b * std::exp(b * smooth_factor)) /
+           ((std::exp(a * smooth_factor) + std::exp(b * smooth_factor)));
+  }
+
+  /**
+   * @brief Return the distance, the center point, and the normal between the current shape and the shape given in the argument. The center point is the point where both shapes are at the same distance from each other. The normal is defined using the closest surface point on the two shapes. By default, the function does not calculate the distance between the two shapes if their bounding boxes are not in contact. This behavior can be modified using the appropriate parameter.
+   * @param shape The shape with which the distance is evaluated
+   * @param cell The cell that is likely to contain the evaluation point
+   * @param candidate_points This is the initial guess points used in the calculation.
+   * @param precision This is the precision of the distance between the two shapes.
+   * @param exact_distance_outside_of_contact This is a boolean to force the exact distance evaluation if the shapes are not in contact.
+   */
+  virtual std::tuple<double, Tensor<1, dim>, Point<dim>>
+  distance_to_shape_with_cell_guess(
```

Comment:
I could not understand the purpose of this function from only the name. Is there a clearer name that could be used? I understand that it does a lot, so finding a good name isn't easy.

Reply:
Well, it is named like this as its primary function is to return the distance of the current shape with the shape given in the argument.